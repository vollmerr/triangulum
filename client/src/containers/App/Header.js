import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
  <ul style={{ listStyleType: 'none', padding: 0 }}>
    <li>
      <Link to="/">home</Link>
    </li>
    <li>
      <Link to="/graph">graph</Link>
    </li>
    <li>
      <Link to="/history">history</Link>
    </li>
  </ul>
);

export default Header;
