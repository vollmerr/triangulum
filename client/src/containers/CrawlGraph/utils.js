/**
 * Stratify function
 *
 * Takes a flat array of node objects containing url and parent url fields, and processes 
 * it into a hierarchical structure, specifically an array containing a single object 
 * with a children array and so on. This mimics d3's stratify function, although is designed 
 * to create the specific formatting requried by react-d3-tree. Additionally, each node
 * object is also given a nodeSvgShape field specified by react-d3-tree for individual node
 * styling. More can be read here: https://github.com/bkrem/react-d3-tree#usage
 *
 * Params: nodes - flat array of nodes objects
 * Returns: data - array with a single object containing children objects
 */
export default function stratify(nodes) {
  
  let data = [];
  findTargetPath(nodes);
  seed(nodes[0], data);
  
  if(nodes.length > 1) {
    for(let i = 0; i < nodes.length; i++) 
      buildHierarchy(nodes[i], data[0]);
  }
  
  addStyling(data[0]);
  return data;

  // Finds target path from the nodes
  function findTargetPath(nodes) {    
    addTargetPathVar(nodes);
    if(nodes[nodes.length - 1].targetFound === 1)
      addTargetPath(nodes, nodes.length - 1, 0);

    // Assigns each object a "targetPath" key and value
    function addTargetPathVar(nodes) {
      for (let i = 0; i < nodes.length; i++)
        nodes[i].targetPath = 0;
    }
    
    // Recursivley determines target path by linear search
    function addTargetPath(nodes, chiIdx, iter) {
      nodes[chiIdx].targetPath = 1;
      if(nodes[chiIdx].parent != null) {
        for(let i = 0; i < nodes.length; i++) {
          if(nodes[chiIdx].parent === nodes[i].url)
            addTargetPath(nodes, i, iter);
        }
      }
    }
  }
  
  // Adds root node to hierarchy
  function seed(node, data) {
    addNodeObj(node, data);
  }

  // Adds new node object to end of array
  function addNodeObj(node, arr) {
    if(node.title) {
      arr.push({ 
        name: node.title,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
    else {
      arr.push({ 
        name: node.url,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
  }

  // Recursively builds hierarchy 
  function buildHierarchy(node, dataObj) {
    if(dataObj.data.url === node.parent) 
      insertChild(node, dataObj);
    else if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        buildHierarchy(node, dataObj.children[i]);
    }        
  }

  // Inserts child node obj onto current node
  function insertChild (node, dataObj) {
    if(!dataObj.children) 
      dataObj.children = [];
    addNodeObj(node, dataObj.children);
  }

  // Recursively goes through hierarchy adding styling object
  function addStyling(dataObj) {
    styling(dataObj);
    if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        addStyling(dataObj.children[i]);
    }        
  }

  // Add nodeSvgShape object to nodes for individual styling
  function styling(dataObj) {

    // Add empty objects
    dataObj.nodeSvgShape = {};
    dataObj.nodeSvgShape.shapeProps = {};

    // Set size
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.r = 20;
    else 
      dataObj.nodeSvgShape.shapeProps.r = 10;

    // Set fill
    if(dataObj.children) {
      if(dataObj.data.targetPath)
        dataObj.nodeSvgShape.shapeProps.fill = "#c0f7cc";
      else
        dataObj.nodeSvgShape.shapeProps.fill = "#e8e8e8";
    } 
    else 
      dataObj.nodeSvgShape.shapeProps.fill = "#fff";

    // Set outline color
    if(dataObj.data.targetPath) {
      if(dataObj.data.targetFound)
        dataObj.nodeSvgShape.shapeProps.stroke = "#dc3545";
      else
        dataObj.nodeSvgShape.shapeProps.stroke = "#28a745";
    }
    else
      dataObj.nodeSvgShape.shapeProps.stroke = "#b7b7b7";

    // Set outline width
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 5; 
    else 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 3;

    // Set cursor when over element to pointer
    dataObj.nodeSvgShape.shapeProps.cursor = 'pointer';
  }
}
