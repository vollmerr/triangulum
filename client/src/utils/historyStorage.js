import uuid from 'uuid/v4';

// NOTE: all the things are stored as strings in localStorage...

export const getItems = () => JSON.parse(localStorage.getItem('crawlHistory'));

export const getItem = ({ id }) => getItems()[id];

export const getCurrent = () => {
  const id = localStorage.getItem('crawlId');
  return id ? getItem({ id }) : {};
};

export const setCurrent = ({ id }) => localStorage.setItem('crawlId', id);

export const newItem = () => {
  const id = uuid();
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      id,
      created: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', id);
  return id;
};

export const updateItem = ({ id, ...item }) => {
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      ...getItem({ id }),
      ...item,
      updated: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
};

export const reset = () => {
  const crawlHistory = JSON.stringify({});
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', '');
};

export const init = () => {
  if (!localStorage.getItem('crawlHistory')) {
    reset();
  }
};
