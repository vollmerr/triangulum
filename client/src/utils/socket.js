const host = process.env.REACT_APP_URL;

let ws = new WebSocket(host); // eslint-disable-line

ws.init = () => {
  ws = new WebSocket(host);
};

ws.onclose = ws.init;

export default ws;
