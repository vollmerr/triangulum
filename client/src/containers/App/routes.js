import CrawlForm from '../CrawlForm';
import CrawlGraph from '../CrawlGraph';
import CrawlHistory from '../CrawlHistory';

const routes = {
  form: {
    key: 'form',
    exact: true,
    name: 'New',
    path: '/',
    icon: 'plus',
    component: CrawlForm,
  },
  graph: {
    key: 'graph',
    name: 'Graph',
    path: '/graph',
    icon: 'chart area',
    component: CrawlGraph,
  },
  history: {
    key: 'history',
    name: 'History',
    path: '/history',
    icon: 'history',
    component: CrawlHistory,
  },
};

export default Object.values(routes);
