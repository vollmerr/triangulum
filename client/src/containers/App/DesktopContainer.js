import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import {
  Container,
  Icon,
  Menu,
  Responsive,
  Segment,
} from 'semantic-ui-react';

import routes from './routes';

export const Logo = () => (
  <Menu.Item as={Link} to={'/'}>
    <Icon name={'sitemap'} style={{ transform: 'rotate(0.5turn)' }} />
    <span>riangulum</span>
  </Menu.Item>
);

export const Links = ({ path }) => (
  routes.map(route => (
    <Menu.Item
      key={route.key}
      icon={route.icon}
      content={route.name}
      as={Link}
      to={route.path}
      active={path === route.path}
    />
  ))
);

export class DesktopContainer extends Component {
  state = {}

  render() {
    const { children, location } = this.props;

    return (
      <Responsive minWidth={Responsive.onlyTablet.minWidth} id={'desktop-view'}>

        <Segment
          inverted
          textAlign='center'
          style={{ minHeight: 40, padding: '1em 0em', margin: '0 0 2em 0' }}
          color={'violet'}
          vertical
        >
          <Container>
            <Menu
              fixed={'top'}
              inverted
              pointing
              secondary
              color={'violet'}
            >
              <Logo />
              <Menu.Item position={'right'} />
              <Links path={location.pathname} />
            </Menu>
          </Container>
        </Segment>

        {children}
      </Responsive>
    );
  }
}

export default withRouter(DesktopContainer);
