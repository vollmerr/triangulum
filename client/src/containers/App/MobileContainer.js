import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import {
  Container,
  Icon,
  Menu,
  Responsive,
  Segment,
  Sidebar,
} from 'semantic-ui-react';

import routes from './routes';

const Links = ({ path, onClick }) => (
  routes.map(route => (
    <Menu.Item
      key={route.key}
      icon={route.icon}
      content={route.name}
      as={Link}
      to={route.path}
      active={path === route.path}
      onClick={onClick}
    />
  ))
);

class MobileContainer extends Component {
  state = {}

  closeMenu = () => {
    const { isMenuOpen } = this.state;
    if (isMenuOpen) {
      this.setState({ isMenuOpen: false });
    }
  }

  toggleMenu = () => {
    const { isMenuOpen } = this.state;
    this.setState({ isMenuOpen: !isMenuOpen });
  }

  render() {
    const { children, location } = this.props;
    const { isMenuOpen } = this.state;

    return (
      <Responsive maxWidth={Responsive.onlyMobile.maxWidth} id={'mobile-view'}>
        <Sidebar.Pushable>
          <Sidebar as={Menu} animation='overlay' vertical visible={isMenuOpen} direction={'right'}>
            <Links path={location.pathname} onClick={this.closeMenu} />
          </Sidebar>

          <Sidebar.Pusher
            dimmed={isMenuOpen}
            onClick={this.closeMenu}
          >
            <Segment
              inverted
              textAlign='center'
              style={{ minHeight: 40, padding: '1em 0em', margin: '0 0 1.5em 0' }}
              color={'violet'}
              vertical
            >
              <Container>
                <Menu inverted pointing secondary fixed='top'>
                  <Menu.Item
                    icon={'sitemap'}
                    style={{ transform: 'rotate(0.5turn)' }}
                    as={Link}
                    to={'/'}
                  />
                  <Menu.Item onClick={this.toggleMenu} position='right'>
                    <Icon name='sidebar' />
                  </Menu.Item>
                </Menu>
              </Container>
            </Segment>

            {children}
          </Sidebar.Pusher>
        </Sidebar.Pushable>
      </Responsive>
    );
  }
}

export default withRouter(MobileContainer);
