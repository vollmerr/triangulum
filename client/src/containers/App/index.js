import React from 'react';
import { HashRouter, Route } from 'react-router-dom';

import * as historyStorage from '../../utils/historyStorage';
import routes from './routes';
import MobileContainer from './MobileContainer';
import DesktopContainer from './DesktopContainer';

// load our websocket
import '../../utils/socket';

/**
 * main application container that handles:
 *  - routing bewteen containers
 *  - tranfering data to pages
 */
class App extends React.PureComponent {
  state = {
    data: {
      response: {
        nodes: [],
      },
    },
  }

  componentDidMount() {
    this.initHistory();
  }

  initHistory = () => {
    historyStorage.init();
    const data = historyStorage.getCurrent();
    this.setState({ data });
  }

  getRoutes = () => {
    const { data } = this.state;
    const props = {
      data,
      updateData: this.updateData,
    };

    return routes.map(route => (
      <Route
        key={route.key}
        exact={route.exact}
        path={route.path}
        component={() => <route.component {...props} />}
      />
    ));
  }

  updateData = ({ data }) => {
    historyStorage.setCurrent({ id: data.id });
    this.setState({ data });
  };

  render() {
    return (
      <HashRouter>
        <div>
          <MobileContainer>
            {this.getRoutes()}
          </MobileContainer>
          <DesktopContainer>
            {this.getRoutes()}
          </DesktopContainer>
        </div>
      </HashRouter>
    );
  }
}

export default App;
