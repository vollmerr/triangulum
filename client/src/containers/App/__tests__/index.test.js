import React from 'react';
import { shallow } from 'enzyme';

import * as historyStorage from '../../../utils/historyStorage';

import App from '../index';

const data = { k1: 'v1' };

let wrapper;
let instance;
describe('App', () => {
  beforeEach(() => {
    wrapper = shallow(<App />);
    instance = wrapper.instance();
  });

  describe('componentDidMount', () => {
    it('should initalize the history on mount if it doesn`t exist', () => {
      instance.componentDidMount();
      expect(historyStorage.getItems()).toEqual({});
    });
  });

  describe('updateData', () => {
    it('should update the local data', () => {
      instance.updateData({ data });
      expect(instance.state.data).toEqual(data);
    });
  });

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
