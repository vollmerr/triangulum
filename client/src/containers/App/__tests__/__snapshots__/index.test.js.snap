// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render should render correctly 1`] = `
<HashRouter>
  <Container
    text={true}
  >
    <Header />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      path="/graph"
    />
    <Route
      component={[Function]}
      path="/history"
    />
  </Container>
</HashRouter>
`;
